// ng2-intl/src/pipes/date.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Pipe, ChangeDetectorRef } from '@angular/core';
import { IntlService, FormatService } from '../services';
import { AbstractI18nPipe } from './abstractI18n';
var FormattedDatePipe = /** @class */ (function (_super) {
    __extends(FormattedDatePipe, _super);
    function FormattedDatePipe(intlService, _ref, formatService) {
        return _super.call(this, intlService, _ref, formatService) || this;
    }
    FormattedDatePipe.prototype.isValidQuery = function (query) {
        return query && (query instanceof Date || isFinite(query));
    };
    FormattedDatePipe.prototype.updateValue = function (query, options) {
        this.value = this.formatService.formatDate(query, options);
        this.lastKey = query;
        this._ref.markForCheck();
    };
    FormattedDatePipe = __decorate([
        Pipe({
            name: 'formattedDate',
            pure: false // required to update the value when the promise is resolved
        }),
        __metadata("design:paramtypes", [IntlService, ChangeDetectorRef, FormatService])
    ], FormattedDatePipe);
    return FormattedDatePipe;
}(AbstractI18nPipe));
export { FormattedDatePipe };
