// ng2-intl/src/components/abstractI18n.js
var AbstractI18nComponent = /** @class */ (function () {
    function AbstractI18nComponent(intlService, formatService) {
        this.intlService = intlService;
        this.formatService = formatService;
    }
    AbstractI18nComponent.prototype.ngOnInit = function () {
        var _this = this;
        // set the value
        this.updateValue();
        // if there is a subscription to onLangChange, clean it
        this._dispose();
        // subscribe to onLangChange event, in case the language changes
        if (!this.onLangChange) {
            this.onLangChange = this.intlService.onLangChange.subscribe(function (event) {
                _this.updateValue();
            });
        }
    };
    /**
    * Clean any existing subscription to change events
    * @private
    */
    AbstractI18nComponent.prototype._dispose = function () {
        if (typeof this.onLangChange !== 'undefined') {
            this.onLangChange.unsubscribe();
            this.onLangChange = undefined;
        }
    };
    AbstractI18nComponent.prototype.ngOnDestroy = function () {
        this._dispose();
    };
    return AbstractI18nComponent;
}());
export { AbstractI18nComponent };
